### PR TITLE
fix: running element page counter incorrect

### PR DESCRIPTION
**Pull request by EnochGao**
#1325 
I'm not sure if this modification is elegant, but the result seems to be correct. I'm not sure how to handle the total number of pages

### CODE REVIEW

**Comment by MurakamiShinyu** [comment]
Thank you for the fix! I will check tomorrow.

**Comment by EnochGao** [comment]
and target-counter in running element equally incorrect🙋🙋🙋

File: packages/core/src/vivliostyle/counters.ts [hunk 1]
```diff
@@ -166,8 +166,12 @@ class CounterResolver implements CssCascade.CounterResolver {
...
     );
 
-    const arrayFormat = (arr) => {
-      return format(arr[0]);
+    const arrayFormat = (arr: number[]) => {
+      let value = arr[0];
+      if (arr[1]) {
+        value = arr[1];
+      }
+      return format(value);
```
Comment (MurakamiShinyu) [inline packages/core/src/vivliostyle/counters.ts]:
```suggestion
      return format(arr[arr.length - 1]);
```

`value = arr[1]` is not good when `counter-reset: page` happens multiple times, for example: 

      @page :nth(2) {
        counter-reset: page 20;
      }
      @page :nth(3) {
        counter-reset: page 30;
      }

**Comment by MurakamiShinyu** [comment]
> I'm not sure how to handle the total number of pages

I found the problem in `CounterStore.finishLastPage()` in `counters.ts`:

https://github.com/vivliostyle/vivliostyle.js/blob/126436b8beb6150e6db70318251499c836c3a650/packages/core/src/vivliostyle/counters.ts#L877-L886

In this code, the "pages" counter value is obtained from the "page" counter value at the last page:

```ts
    const pages = this.currentPageCounters["page"][0];
```

It doesn't work when `counter-reset: page` is used for the same reason as `return format(arr[0]);` in `getPageCounterVal()`. However, there is a further problem.

In the [CSS Paged Media](https://drafts.csswg.org/css-page-3/#page-based-counters) spec, the definition of the `pages` counter is "Its value is always the total number of pages in the document" and not same as the "page" counter value at the last page. So the "pages" counter value should be obtained from the number of generated pages.

Suggested change:

```ts
    const pages = viewport.contentContainer.childElementCount;
```

**Comment by MurakamiShinyu** [comment]
With the last commit https://github.com/vivliostyle/vivliostyle.js/pull/1327/commits/f5083bb6d970d24e663aae16877872f23721c8a6 the test result seems good except the last page with "??" remaining.

https://vivliostyle-7iz81mer5-vivliostyle.vercel.app/#src=https://gist.githubusercontent.com/MurakamiShinyu/7616ab180e72e38addafc1a7ea44d8d0/raw/issue-1325-running-element-target-counter.html

Screenshot 2024-05-21 21 50 07

**Comment by MurakamiShinyu** [comment]
@EnochGao Thanks for the fix and the test case "counter_in_running.html".

The fix seems almost good, but I found a problem that it does not work when a page-based 
counter other than "page" is used. Example: https://gist.githubusercontent.com/MurakamiShinyu/7616ab180e72e38addafc1a7ea44d8d0/raw/counter_in_running_custom_page.html

**Comment by MurakamiShinyu** [comment]
Thanks. The fix 72fa5bc is good for custom page-based counter, but still has a problem with element-based counter.

I made a test case which tests element-based counter in addition to custom page-based counter:

https://gist.githubusercontent.com/MurakamiShinyu/7616ab180e72e38addafc1a7ea44d8d0/raw/counter_in_running_custom_page2.html

Test: https://vivliostyle-git-fork-enochgao-fix-counter-vivliostyle.vercel.app/#src=https://gist.githubusercontent.com/MurakamiShinyu/7616ab180e72e38addafc1a7ea44d8d0/raw/counter_in_running_custom_page2.html

The "(Section 11)" in the page header is wrong. It should be "(Section 2)".

-----

I found this problem is caused by the following code in `finishLastPage()`:

```ts
      const transformedId = node.getAttribute(TARGET_PAGE_COUNTER_ID_ATTR);
      const expr = this.targetPageCounterExprs.find(
        (o) => o.transformedId === transformedId,
      );
```

There are multiple target counter expressions with the same `transformedId` in `this.targetPageCounterExprs` and the wrong one is selected. I think we should use `key` instead of `transformedId` to identify the target counter expression. So the code should be changed to:

```ts
      const key = node.getAttribute(TARGET_PAGE_COUNTER_ATTR);
      const expr = this.targetPageCounterExprs.find((o) => o.expr.key === key);
```

If this change is ok, the `TARGET_PAGE_COUNTER_ID_ATTR` attribute is no longer needed and can be removed.

Also the naming `TARGET_PAGE_COUNTER_ATTR` is no longer appropriate because it is used for not only page-based counters but also element-based counters. I suggest renaming it to `TARGET_COUNTER_ATTR` and `TARGET_PAGE_COUNTER_IN_RUNNING_ATTR` to `TARGET_COUNTER_IN_RUNNING_ATTR`, and change their values to `"data-vivliostyle-target-counter"` and `"data-vivliostyle-target-counter-in-running"` respectively.

Please check the proposed changes in the following diff:
https://gist.github.com/MurakamiShinyu/d175de0ddb9175e3f2ba711735aa0e17

**Comment by EnochGao** [comment]
I see. Thanks for the correction.

File: packages/core/test/files/counter_in_running.html [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+          placeholder placeholder placeholder placeholder placeholder
+          placeholder placeholder placeholder placeholder placeholder
+          placeholder placeholder placeholder placeholder placeholder
+          placeholder placeholder placeholder placeholder placeholder
+          placeholder placeholder placeholder placeholder placeholder
+          placeholder placeholder placeholder placeholder placeholder
+        </p>
+
+        <div class="reset-page">reset page</div>
+
```
Comment (MurakamiShinyu) [inline packages/core/test/files/counter_in_running.html]: This `<div class="reset-page">reset page</div>` element is not used for resetting page counter. So I think this should be removed.



```suggestion
```

Comment (EnochGao) [reply packages/core/test/files/counter_in_running.html]: ok

**Review by MurakamiShinyu — APPROVED** [review]
Thanks! LGTM 💯

**Comment by EnochGao** [comment]
> Thanks! LGTM 💯

Thanks for the correction. Made me know more about this place

**Pull request merged**